src/main/java/org/apache/datasketches/filters/bloomfilter/BitArray.java

```diff
@@ -0,0 +1,219 @@
+  private long[] data_;
+
+  // creates an array of a given size
+  BitArray(final long numBits) {
+    if (numBits <= 0) {
+      throw new SketchesArgumentException("Number of bits must be strictly positive. Found: " + numBits);
+    }
+    if (numBits > MAX_BITS) {
+      throw new SketchesArgumentException("Number of bits may not exceed " + MAX_BITS + ". Found: " + numBits);
+    }
+
+    final int numLongs = (int) Math.ceil(numBits / 64.0);
+    numBitsSet_ = 0;
+    data_ = new long[numLongs];
+  }
+
+  // uses the provided array
+  BitArray(final long[] data) {
+    data_ = data;
+    numBitsSet_ = 0;
+    for (long val : data) {
+      numBitsSet_ += Long.bitCount(val);
+    }
+  }
+
+  // reads a serialized image, but the BitArray is not fully self-describing so requires
+  // a flag to indicate whether the array is empty
+  static BitArray heapify(final Memory mem, final boolean isEmpty) {
+    final int numLongs = mem.getInt(0);
+    if (numLongs < 0) {
+      throw new SketchesArgumentException("Possible corruption: Must have strictly positive array size. Found: " + numLongs);
+    }
+    
+    if (isEmpty) {
+      return new BitArray((long) numLongs * Long.SIZE);
+    }
+
+    final long[] data = new long[numLongs];
+    mem.getLongArray(DATA_OFFSET, data, 0, numLongs);
+    return new BitArray(data);
+  }
+
+  boolean isEmpty() {
+    return numBitsSet_ == 0;
+  }
+
+  // queries a single bit in the array
+  boolean getBit(final long index) {
+    return (data_[(int) index >>> 6] & (1L << index)) != 0 ? true : false;
+  }
+
+  // sets a single bit in the array without querying, meaning the method
+  // cannot properly track the number of bits set. 
+  // instead changes numBitsSet_ to indicate that the value is unreliable
+  void setBit(final long index) {
+    data_[(int) index >>> 6] |= 1L << index;
+    numBitsSet_ = -1; // use as a dirty flag
+  }
+
+  // returns existing value of bit
+  boolean getAndSetBit(final long index) {
+    final int offset = (int) index >>> 6;
+    final long mask = 1L << index;
+    if ((data_[offset] & mask) != 0) {
+      return true; // already seen
+    } else {
+      data_[offset] |= mask;
+      if (numBitsSet_ != -1) { ++numBitsSet_; }
+      return false; // new set
+    }
+  }
+
+  // may need to recompute value:
+  // O(1) if only getAndSetBit() has been used
+  // O(data_.length) if setBit() has ever been used
+  long getNumBitsSet() {
+    if (numBitsSet_ == -1) {
+      numBitsSet_ = 0;
+      for (long val : data_) {
+        numBitsSet_ += Long.bitCount(val);
+      }  
+    }
+    return numBitsSet_;
+  }
+
+  long getCapacity() { return (long) data_.length * Long.SIZE; }
+
+  int getArrayLength() { return data_.length; }
+
+  // applies logical OR
+  void union(final BitArray other) {
+    if (data_.length != other.data_.length) {
+      throw new SketchesArgumentException("Cannot union bit arrays with unequal lengths");
+    }
+
+    long numBitsSet = 0;
```

Comment:
why not use the existing member variable? would it be slower?

Reply:
bias towards not modifying the member variable constantly, but i think that's silly of me